lib/backend/k8s/k8s.go

```diff
@@ -130,18 +135,116 @@ func NewKubeClient(kc *apiconfig.KubeConfig) (api.Client, error) {
+		apiv2.KindBGPConfiguration,
+		resources.NewBGPConfigClient(cs, crdClientV1),
+	)
+	kubeClient.registerResourceClient(
+		reflect.TypeOf(model.ResourceKey{}),
+		reflect.TypeOf(model.ResourceListOptions{}),
+		apiv2.KindFelixConfiguration,
+		resources.NewFelixConfigClient(cs, crdClientV1),
+	)
+	kubeClient.registerResourceClient(
+		reflect.TypeOf(model.ResourceKey{}),
+		reflect.TypeOf(model.ResourceListOptions{}),
+		apiv2.KindClusterInformation,
+		resources.NewClusterInfoClient(cs, crdClientV1),
+	)
+	kubeClient.registerResourceClient(
+		reflect.TypeOf(model.ResourceKey{}),
+		reflect.TypeOf(model.ResourceListOptions{}),
+		apiv2.KindNode,
+		resources.NewNodeClient(cs),
+	)
+	kubeClient.registerResourceClient(
+		reflect.TypeOf(model.ResourceKey{}),
+		reflect.TypeOf(model.ResourceListOptions{}),
+		apiv2.KindProfile,
+		resources.NewProfileClient(cs),
+	)
+	kubeClient.registerResourceClient(
+		reflect.TypeOf(model.BlockAffinityKey{}),
+		reflect.TypeOf(model.BlockAffinityListOptions{}),
+		"",
+		resources.NewAffinityBlockClient(cs),
+	)
 
 	return kubeClient, nil
 }
 
+// registerResourceClient registers a specific resource client with the associated
+// key and list types (and for v2 resources with the resource kind - since these share
+// a common key and list type).
+func (c *KubeClient) registerResourceClient(keyType, listType reflect.Type, resourceKind string, client resources.K8sResourceClient) {
+	if keyType == resourceKeyType {
+		c.clientsByResourceKind[resourceKind] = client
+	} else {
+		c.clientsByKeyType[keyType] = client
+		c.clientsByListType[listType] = client
+	}
+}
+
+// getResourceClientFromKey returns the appropriate resource client for the v2 resource kind.
+func (c *KubeClient) getResourceClientFromResourceKind(kind string) (resources.K8sResourceClient, error) {
+	client := c.clientsByResourceKind[kind]
+	if client == nil {
+		return nil, fmt.Errorf("Unhandled resource type %v", kind)
+	}
+	return client, nil
+}
+
+// getResourceClientFromKey returns the appropriate resource client for the key.
+func (c *KubeClient) getResourceClientFromKey(key model.Key) (resources.K8sResourceClient, error) {
+	var client resources.K8sResourceClient
+	kt := reflect.TypeOf(key)
+	if kt == resourceKeyType {
+		client = c.clientsByResourceKind[key.(model.ResourceKey).Kind]
+	} else {
+		client = c.clientsByKeyType[kt]
+	}
+	if client == nil {
+		return nil, fmt.Errorf("Unhandled resource type %v", kt)
+	}
+	return client, nil
+}
+
+// getResourceClientFromList returns the appropriate resource client for the list.
+func (c *KubeClient) getResourceClientFromList(list model.ListInterface) (resources.K8sResourceClient, error) {
+	var client resources.K8sResourceClient
+	lt := reflect.TypeOf(list)
+	if lt == resourceListType {
+		client = c.clientsByResourceKind[list.(model.ResourceListOptions).Kind]
+	} else {
+		client = c.clientsByListType[lt]
+	}
+	if client == nil {
+		return nil, fmt.Errorf("Unhandled resource type %v", lt)
```

Comment:
Given this error is propagated to the user, I think this should by an UnsupportedOperation or whatever, with a more user-facing message.

Fine to punt post-beta so long as there is an issue tracking it within the v3.0 milestone.

Reply:
Oooh yes it will be - i'll fix it up!

Reply:
Actually I think I'll fix up to not return an error, but just return nil and the calling code can tailor the error appropriately.